plugins/notifications-backend/src/service/router.ts

```diff
@@ -8,24 +20,111 @@ import { Context, OpenAPIBackend, Request } from 'openapi-backend';
 import { Logger } from 'winston';
 
 import { Paths } from '../openapi';
+import {
+  notificationsCreatePermission,
+  notificationsPermissions,
+  notificationsReadPermission,
+  notificationsSetReadPermission,
+} from '../permissions';
 import { initDB } from './db';
 import {
   createNotification,
   getNotifications,
   getNotificationsCount,
   setRead,
 } from './handlers';
+import { DefaultServiceUser } from './types';
 
 interface RouterOptions {
   logger: Logger;
   dbConfig: Config;
   catalogClient: CatalogClient;
+  identity: IdentityApi;
+  permissions: PermissionEvaluator;
+  tokenManager: TokenManager;
+  notificationsServiceToServiceAuthEnabled: boolean;
 }
 
 export async function createRouter(
   options: RouterOptions,
 ): Promise<express.Router> {
-  const { logger, dbConfig, catalogClient } = options;
+  const {
+    logger,
+    dbConfig,
+    catalogClient,
+    identity,
+    permissions,
+    tokenManager,
+    notificationsServiceToServiceAuthEnabled,
+  } = options;
+
+  /*
+   * User's entity must be present in the catalog.
+   */
+  const getLoggedInUser = async (request: express.Request): Promise<string> => {
+    const identityResponse = await identity.getIdentity({ request });
+
+    // To properly set identity, see packages/backend/src/plugins/auth.ts or https://backstage.io/docs/auth/identity-resolver
+    if (identityResponse) {
+      // user:default/guest
+      let author = identityResponse?.identity.userEntityRef;
+      if (author) {
+        if (author.startsWith('user:')) {
+          author = author.slice('user:'.length);
+        }
+      } else {
+        throw new AuthenticationError(
+          'Missing valid authentication data or the user is not in the Catalog.',
+        );
+      }
+      return author;
+    }
+
+    if (notificationsServiceToServiceAuthEnabled) {
+      logger.info(
+        'Identity not found for the request, trying to authorize service-to-service.',
+      );
+      const token = getBearerTokenFromAuthorizationHeader(
+        request.header('authorization'),
+      );
+      if (!token) {
+        throw new AuthenticationError(
+          'Missing authorization token in the request.',
+        );
+      }
+      await tokenManager.authenticate(token);
+    }
+
+    logger.info(
+      `Identity not found for the request. Since the service to service authentication is disabled, using default service user "${DefaultServiceUser}" instead.`,
+    );
+    return DefaultServiceUser;
+  };
+
+  const checkPermission = async (
+    request: express.Request,
+    permission: BasicPermission,
+    loggedInUser: string,
+  ) => {
+    const token = getBearerTokenFromAuthorizationHeader(
+      request.header('authorization'),
+    );
+    const decision = (
+      await permissions.authorize([{ permission }], {
```

Comment:
how does it know for which user to check permissions? did u test it?

Reply:
Yes. It's about the "injected" permission policies, out of our scope.